File: spark/src/test/scala/geotrellis/spark/io/avro/TileCodecsSpec.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark.io.avro

import geotrellis.spark.io.avro.codecs.TileCodecs
import org.scalatest._
import TileCodecs._
import geotrellis.raster._
import geotrellis.spark.io.avro.AvroTools._

class TileCodecsSpec extends FunSpec with Matchers with AvroTools  {
  describe("TileCodecs") {
    it("encodes ByteArrayTile"){
      roundTripWithNoDataCheck(ByteArrayTile.fill(127,10,15))
    }
    it("encodes UByteArrayTile"){
      roundTripWithNoDataCheck(UByteArrayTile.fill(127,10,15))
    }
    it("encodes ShortArrayTile"){
      roundTripWithNoDataCheck(ShortArrayTile.fill(45,10,15))
    }
    it("encodes UShortArrayTile"){
      roundTripWithNoDataCheck(UShortArrayTile.fill(45,10,15))
    }
    it("encodes IntArrayTile"){
      roundTripWithNoDataCheck(IntArrayTile.fill(45,10,15))
    }
    it("encodes FloatArrayTile"){
      roundTripWithNoDataCheck(FloatArrayTile.fill(532.4f,10,15))
    }
    it("encodes DoubleArrayTile"){
      roundTripWithNoDataCheck(DoubleArrayTile.fill(53232322.4,10,15))
    }

    it("encodes ArrayMultibandTile"){
      val tiles= for (i <- 0 to 3) yield DoubleArrayTile.fill(53232322.4,10,15)
      val thing = ArrayMultibandTile(tiles): MultibandTile
      roundTripWithNoDataCheck(thing)
//      val bytes = AvroEncoder.toBinary(thing)
//      val fromBytes = AvroEncoder.fromBinary[MultibandTile](bytes)
//      fromBytes should be equals (thing)
    }
  }

  describe("No NoData TileCodecs") {
    it("encodes ByteArrayTile"){
      roundTripWithNoDataCheck(ByteArrayTile.fill(127,10,15, ByteCellType))
    }
    it("encodes UByteArrayTile"){
      roundTripWithNoDataCheck(UByteArrayTile.fill(127,10,15, UByteCellType))
    }
    it("encodes ShortArrayTile"){
      roundTripWithNoDataCheck(ShortArrayTile.fill(45,10,15, ShortCellType))
    }
    it("encodes UShortArrayTile"){
      roundTripWithNoDataCheck(UShortArrayTile.fill(45,10,15, UShortCellType))
    }
    it("encodes IntArrayTile"){
      roundTripWithNoDataCheck(IntArrayTile.fill(45,10,15, IntCellType))
    }
    it("encodes FloatArrayTile"){
      roundTripWithNoDataCheck(FloatArrayTile.fill(532.4f,10,15, FloatCellType))
    }
    it("encodes DoubleArrayTile"){
      roundTripWithNoDataCheck(DoubleArrayTile.fill(53232322.4,10,15, DoubleCellType))
    }
    it("encodes ArrayMultibandTile"){
      val tiles= for (i <- 0 to 3) yield DoubleArrayTile.fill(53232322.4,10,15, DoubleCellType)
      val thing = ArrayMultibandTile(tiles): MultibandTile
      roundTripWithNoDataCheck(thing)
    }
  }

  describe("UserDefined TileCodecs") {
    it("encodes ByteArrayTile"){
      roundTripWithNoDataCheck(ByteArrayTile.fill(127,10,15, ByteUserDefinedNoDataCellType(123)))
    }
    it("encodes UByteArrayTile"){
      roundTripWithNoDataCheck(UByteArrayTile.fill(127,10,15, UByteUserDefinedNoDataCellType(123)))
    }
    it("encodes ShortArrayTile"){
      roundTripWithNoDataCheck(ShortArrayTile.fill(45,10,15, ShortUserDefinedNoDataCellType(123)))
    }
    it("encodes UShortArrayTile"){
      roundTripWithNoDataCheck(UShortArrayTile.fill(45,10,15, UShortUserDefinedNoDataCellType(123)))
    }
    it("encodes IntArrayTile"){
      roundTripWithNoDataCheck(IntArrayTile.fill(45,10,15, IntUserDefinedNoDataCellType(123)))
    }
    it("encodes FloatArrayTile"){
      roundTripWithNoDataCheck(FloatArrayTile.fill(532.4f,10,15, FloatUserDefinedNoDataCellType(2.2F)))
    }
    it("encodes DoubleArrayTile"){
      roundTripWithNoDataCheck(DoubleArrayTile.fill(53232322.4,10,15, DoubleUserDefinedNoDataCellType(2.2)))
    }
    it("encodes ArrayMultibandTile"){
      val tiles= for (i <- 0 to 3) yield DoubleArrayTile.fill(53232322.4,10,15, DoubleUserDefinedNoDataCellType(42.23))
      val thing = ArrayMultibandTile(tiles): MultibandTile
      roundTripWithNoDataCheck(thing)
    }
  }
}

File: geotrellis-python/tests/geotrellis/spark/io/avro/TileCodecsSpec.py
from tests.geotrellis.spark.io.avro.AvroTools import AvroTools
from geotrellis.raster.ShortArrayTile import ShortArrayTile
from geotrellis.raster.UShortArrayTile import UShortArrayTile
from geotrellis.raster.IntArrayTile import IntArrayTile
from geotrellis.raster.FloatArrayTile import FloatArrayTile
from geotrellis.raster.DoubleArrayTile import DoubleArrayTile
from geotrellis.raster.ByteArrayTile import ByteArrayTile
from geotrellis.raster.UByteArrayTile import UByteArrayTile
from geotrellis.raster.BitArrayTile import BitArrayTile
from geotrellis.raster.ArrayMultibandTile import ArrayMultibandTile
from spec import Spec

class TileCodecsSpec(AvroTools, Spec):
    class inner_tile_codecs(object):
        "TileCodecs"

        def test_byte(self):
            "encodes ByteArrayTile"
            self.roundTripWithNoDataCheck(ByteArrayTile.fill(127, 10, 15))

        def test_ubyte(self):
            "encodes UByteArrayTile"
            self.roundTripWithNoDataCheck(UByteArrayTile.fill(127, 10, 15))

        def test_short(self):
            "encodes ShortArrayTile"
            self.roundTripWithNoDataCheck(ShortArrayTile.fill(45, 10, 15))

        def test_ushort(self):
            "encodes UShortArrayTile"
            self.roundTripWithNoDataCheck(UShortArrayTile.fill(45, 10, 15))

        def test_int(self):
            "encodes IntArrayTile"
            self.roundTripWithNoDataCheck(IntArrayTile.fill(45, 10, 15))

        def test_float(self):
            "encodes FloatArrayTile"
            self.roundTripWithNoDataCheck(FloatArrayTile.fill(532.4, 10, 15))

        def test_double(self):
            "encodes DoubleArrayTile"
            self.roundTripWithNoDataCheck(DoubleArrayTile.fill(53232322.4, 10, 15))

        def test_multiband(self):
            "encodes ArrayMultibandTile"
            tiles = [DoubleArrayTile.fill(53232322.4, 10, 15)] * 3
            thing = ArrayMultibandTile(tiles)
            self.roundTripWithNoDataCheck(thing)

    class inner_no_nodata_tile_codecs(object):
        "No NoData TileCodecs"

        def test_byte(self):
            "encodes ByteArrayTile"
            self.roundTripWithNoDataCheck(ByteArrayTile.fill(127, 10, 15, ByteCellType))

        def test_ubyte(self):
            "encodes UByteArrayTile"
            self.roundTripWithNoDataCheck(UByteArrayTile.fill(127, 10, 15, UByteCellType))

        def test_short(self):
            "encodes ShortArrayTile"
            self.roundTripWithNoDataCheck(ShortArrayTile.fill(45, 10, 15, ShortCellType))

        def test_ushort(self):
            "encodes UShortArrayTile"
            self.roundTripWithNoDataCheck(UShortArrayTile.fill(45, 10, 15, UShortCellType))

        def test_int(self):
            "encodes IntArrayTile"
            self.roundTripWithNoDataCheck(IntArrayTile.fill(45, 10, 15, IntCellType))

        def test_float(self):
            "encodes FloatArrayTile"
            self.roundTripWithNoDataCheck(FloatArrayTile.fill(532.4, 10, 15, FloatCellType))

        def test_double(self):
            "encodes DoubleArrayTile"
            self.roundTripWithNoDataCheck(DoubleArrayTile.fill(53232322.4, 10, 15, DoubleCellType))

        def test_multiband(self):
            "encodes ArrayMultibandTile"
            tiles = [DoubleArrayTile.fill(53232322.4, 10, 15, DoubleCellType)] * 3
            thing = ArrayMultibandTile(tiles)
            self.roundTripWithNoDataCheck(thing)

    class inner_user_defined_tile_codecs(object):
        "UserDefined TileCodecs"

        def test_byte(self):
            "encodes ByteArrayTile"
            self.roundTripWithNoDataCheck(ByteArrayTile.fill(127, 10, 15, ByteUserDefinedNoDataCellType(123)))

        def test_ubyte(self):
            "encodes UByteArrayTile"
            self.roundTripWithNoDataCheck(UByteArrayTile.fill(127, 10, 15, UByteUserDefinedNoDataCellType(123)))

        def test_short(self):
            "encodes ShortArrayTile"
            self.roundTripWithNoDataCheck(ShortArrayTile.fill(45, 10, 15, ShortUserDefinedNoDataCellType(123)))

        def test_ushort(self):
            "encodes UShortArrayTile"
            self.roundTripWithNoDataCheck(UShortArrayTile.fill(45, 10, 15, UShortUserDefinedNoDataCellType(123)))

        def test_int(self):
            "encodes IntArrayTile"
            self.roundTripWithNoDataCheck(IntArrayTile.fill(45, 10, 15, IntUserDefinedNoDataCellType(123)))

        def test_float(self):
            "encodes FloatArrayTile"
            self.roundTripWithNoDataCheck(FloatArrayTile.fill(532.4, 10, 15, FloatUserDefinedNoDataCellType(2.2)))

        def test_double(self):
            "encodes DoubleArrayTile"
            self.roundTripWithNoDataCheck(DoubleArrayTile.fill(53232322.4, 10, 15, DoubleUserDefinedNoDataCellType(2.2)))

        def test_multiband(self):
            "encodes ArrayMultibandTile"
            tiles = [DoubleArrayTile.fill(53232322.4, 10, 15, DoubleUserDefinedNoDataCellType(42.23))] * 3
            thing = ArrayMultibandTile(tiles)
            self.roundTripWithNoDataCheck(thing)
